design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,241 @@
+
+A command-line switch (defaulting to off) will enable integration between the
+baremetal-operator and the image request controller.
+
+### Implementation Details/Notes/Constraints
+
+When integration is enabled, the controller portion of the baremetal-operator
+will update the `RamDiskImageRequest` with the same name and namespace as the
+`BareMetalHost`, creating it if it doesn't exist, to point to the networkData
+Secret referenced in the `BareMetalHost` Spec. The Architecture will be
+obtained from the Host's hardware profile. (Currently we don't ratify the
+choice of hardware profile until after inspection, but we could move this state
+earlier. In any event, we don't support multiple architectures today, so we
+could use a default value in the interim.)
+
+The ironic provisioner will refuse to report registration complete in
+`ValidateMangementAccess()` until the Ready condition is true in the
+corresponding `RamDiskImageRequest` Status. It will then configure the Node to
+use the ramdisk Image from the URL in the Status. This must be done before
+inspection or cleaning starts, but can be after the Ironic node reaches the
+Manageable state.
+
+The contents of the networkData Secret are generally assumed to be in the same
+format output by OpenStack and consumable by cloud-init. However, for Metal³'s
+purposes this is not a requirement as the data is passed through directly.
+Authors of image customisation controllers are free to interpret the data in
+any way they see fit.
+
+Currently we always supply the deploy ramdisk to ironic in the form of a
+kernel+initrd. Some virtualmedia drivers now allow the image to be in ISO
+format, which is convenient in the sense that many almost-usable images (i.e.
+modulo light customisation) may be supplied by distros in ISO format. There are
+several possible ways of handling this fact:
+
+1. Always return an initrd
+2. Always return an ISO, and restrict building custom images to only
+   virtualmedia drivers (as Ironic itself does when used in an OpenStack cloud)
+3. Add an option in the API to request a particular format
+
+The second option is more reasonable than it sounds, since PXE-booted hosts
+generally must be connected to the provisioning network on an untagged VLAN
+using LACP for any NIC bonding, and on the provisioning network ironic will
+supply an IP address via DHCP. Therefore there should be no need for this
+feature except when using virtualmedia drivers. An API option can always be
+added later if this proves not to be the case.
+
+The `BareMetalHost` should be set as the Owner of the `RamDiskImageRequest` so
+that the request gets deleted when the Host is deleted. This allows controllers
+to use the resource to perform cache management.
+
+Once the image URL is set, it shouldn't be changed. If the user wishes to
+rebuild the ramdisk with newer settings, they can do so by deleting the image
+request. The baremetal-operator should watch the image request and recreate it
```

Comment:
Are the spec field immutable, then? Do they need to be?

Does the BMO actually need to watch the image request, or would it just make one when it sees that it needs one based on host state?